cmd/kubeadm/app/phases/uploadcerts/uploadcerts.go

```diff
@@ -0,0 +1,187 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      KubeadmCertsClusterRoleName,
+			Namespace: metav1.NamespaceSystem,
+		},
+		Rules: []rbac.PolicyRule{
+			rbachelper.NewRule("get").Groups("").Resources("secrets").Names(kubeadmconstants.KubeadmCertsSecret).RuleOrDie(),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return apiclient.CreateOrUpdateRoleBinding(client, &rbac.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      KubeadmCertsClusterRoleName,
+			Namespace: metav1.NamespaceSystem,
+		},
+		RoleRef: rbac.RoleRef{
+			APIGroup: rbac.GroupName,
+			Kind:     "Role",
+			Name:     KubeadmCertsClusterRoleName,
+		},
+		Subjects: []rbac.Subject{
+			{
+				Kind: rbac.GroupKind,
+				Name: kubeadmconstants.NodeBootstrapTokenAuthGroup,
+			},
+		},
+	})
+}
+
+func getOwnerReference(client clientset.Interface, tokenID string) ([]metav1.OwnerReference, error) {
+	secretName := bootstraputil.BootstrapTokenSecretName(tokenID)
+	secret, err := client.CoreV1().Secrets(metav1.NamespaceSystem).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "error to get token reference")
+	}
+
+	gvk := schema.GroupVersionKind{Version: "v1", Kind: "Secret"}
+	ref := metav1.NewControllerRef(secret, gvk)
+	return []metav1.OwnerReference{*ref}, nil
+}
+
+func loadAndEncryptCert(certPath string, key []byte) ([]byte, error) {
+	cert, err := ioutil.ReadFile(certPath)
+	if err != nil {
+		return nil, err
+	}
+	return cryptoutil.EncryptBytes(cert, key)
+}
+
+func certsToUpload(cfg *kubeadmapi.InitConfiguration, certDir string) map[string]string {
+	certs := map[string]string{
+		kubeadmconstants.CACertName:                   path.Join(certDir, kubeadmconstants.CACertName),
+		kubeadmconstants.CAKeyName:                    path.Join(certDir, kubeadmconstants.CAKeyName),
+		kubeadmconstants.FrontProxyCACertName:         path.Join(certDir, kubeadmconstants.FrontProxyCACertName),
+		kubeadmconstants.FrontProxyCAKeyName:          path.Join(certDir, kubeadmconstants.FrontProxyCAKeyName),
+		kubeadmconstants.ServiceAccountPublicKeyName:  path.Join(certDir, kubeadmconstants.ServiceAccountPublicKeyName),
+		kubeadmconstants.ServiceAccountPrivateKeyName: path.Join(certDir, kubeadmconstants.ServiceAccountPrivateKeyName),
+	}
+
+	if cfg.Etcd.External == nil {
+		certs[kubeadmconstants.EtcdCACertName] = path.Join(certDir, kubeadmconstants.EtcdCACertName)
+		certs[kubeadmconstants.EtcdCAKeyName] = path.Join(certDir, kubeadmconstants.EtcdCAKeyName)
+	} else {
+		certs[externalEtcdCA] = cfg.Etcd.External.CAFile
+		certs[externalEtcdCert] = cfg.Etcd.External.CertFile
+		certs[externalEtcdKey] = cfg.Etcd.External.KeyFile
+	}
+	return certs
+}
+
+func getSecretData(cfg *kubeadmapi.InitConfiguration, certsDir string, key []byte) (map[string][]byte, error) {
+	secretData := map[string][]byte{}
+	for certName, certPath := range certsToUpload(cfg, certsDir) {
+		cert, err := loadAndEncryptCert(certPath, key)
```

Comment:
The only case where the CA keys should be missing is if the user has supplied (and pre-deployed) the CA certs, in which case I would expect them to be able to do the same on subsequent contol plane nodes and would not leverage the built in ability to upload certs for kubeadm. I think it is safe to fail here if either the cert or key are missing, otherwise it might cause user confusion why control plane join doesn't work if they download only the CA certs and kubeadm fails to generate any missing server/client certificates.

Reply:
i think i agree with @detiber .
i'm pretty sure we decided to overwrite existing certs if the user requests them on join, but also for the upload we might want to only upload if all of them exist and error if one of them is missing?

@fabriziopandini are you seeing another special case?

> and to silently move on only if certificate authority key is missing
either way, it doesn't hurt us to add some sort of a message.

Reply:
Ok, let's fail fast/with a clear message to the user if we are in the external CA mode (I don't like the idea of silently move on)
@yagonobre I guess we can check this condition in init data